### PR TITLE
refactor: migrate LSP handler deps from di.*() globals to context injection [IDE-1898]

### PR DESCRIPTION
**Pull request by bastiandoetsch**
## Summary
- Migrate all `di.*()` global accessor calls in LSP handler functions to context-based dependency injection
- Extend `di.Dependencies` struct with handler-accessed fields: `Scanner`, `HoverService`, `ScanNotifier`, `ScanPersister`, `FileWatcher`, `ErrorReporter`, `CodeActionService`, `FeatureFlagService`
- Add nil-guarded injections for all new fields in `withContext` middleware
- Add typed context extractors for each new field (`hoverServiceFromContext`, `scannerFromContext`, etc.)
- Remove dead `initializerFromContext` / `installerFromContext` helpers (Installer/Initializer are process-lifecycle deps, not per-request)
- Fix `logError` to call `di.ErrorReporter().CaptureError(err)` (regression restore)
- Add nil guard in `notifyLockedFieldsRejected`; soft-fail in `refreshLdxSyncOnTokenChange`

## PR Stack — Merge Order
```mermaid
flowchart LR
    main(["main"])
    PR1["#1302 PR-1 ← YOU ARE HERE\nDI context injection\n(production code)"]
    PR2["#1303 PR-2\nTDD contract tests"]
    main --> PR1 --> PR2
    style PR1 fill:#ffd700,color:#000
```

## Deferred to later PRs
- `di.TestInit` write-path still mutates package-level globals (PR2 notes this as a known gap; full parallel test safety requires fixing this too)
- Full `t.Parallel()` adoption once the write-path is fixed

## Test plan
- [x] `go build ./...` clean
- [x] `go vet ./...` clean
- [x] `go test ./application/server/...` passes with `-race`
- [x] `go test ./application/di/...` passes
- [x] PR size: 601 changed lines (under 700 limit)

___

### **PR Type**
Enhancement, Bug fix

___

### **Description**
- Migrate LSP handler dependencies to context injection.

- Introduce `must*FromContext` helpers for mandatory dependencies.

- Enhance nil-safety and error handling in config updates.

- Update DI struct, tests, and server startup logic.

___

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>5 files</summary><table>
<tr>
  <td><strong>init.go</strong><dd><code>Update DI Dependencies struct for context injection.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/snyk/snyk-ls/pull/1302/changes#diff-3dfa2a6c1d0ed218f91d033ea2c5e5575f64d80c2575b71fe0605ec4fa41d4f6">+17/-18</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>configuration.go</strong><dd><code>Refactor server handlers to use context DI.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/snyk/snyk-ls/pull/1302/changes#diff-eb3fe6fd1956591838595b08979653ec23c0b0c16f5340e9e89a05c30e921607">+44/-32</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>notification.go</strong><dd><code>Refactor client notification logic to use context.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/snyk/snyk-ls/pull/1302/changes#diff-3ac97f07d29586cd80b9c63b5ac84fa3c0c029c0786ad485ae98d5d143cdaa53">+15/-12</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>server.go</strong><dd><code>Implement context injection wrapper and dependency validation.</code></dd></td>
  <td><a href="https://github.com/snyk/snyk-ls/pull/1302/changes#diff-1b5a200b7dc3f37ed67632d0c2682eeb53762813f96edb15cb7660e076fa6031">+136/-62</a></td>

</tr>

<tr>
  <td><strong>context.go</strong><dd><code>Define new dependency keys for context injection.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/snyk/snyk-ls/pull/1302/changes#diff-5ba4125877fb56856b4bb75f14ff968c77cdee83e09ea4cefd9d24fc2d453923">+2/-4</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>6 files</summary><table>
<tr>
  <td><strong>init_test.go</strong><dd><code>Update tests for DI context injection changes.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/snyk/snyk-ls/pull/1302/changes#diff-fcd7ecadf7c2cc1571bea1e6d6238bb60ffce015cbe3f70cd9b3978a3facd621">+4/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>test_init.go</strong><dd><code>Adjust test initialization for DI context.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/snyk/snyk-ls/pull/1302/changes#diff-fc304a8ceac9f55997a1ed4f30020de933d3ada1a11eacd32c236acdce0cec81">+6/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>codeaction_handlers_test.go</strong><dd><code>Remove obsolete test for nil CodeActionsService.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/snyk/snyk-ls/pull/1302/changes#diff-c7a7efea54d261bb20456900c55c74e98224241d0b1bacdcbfd43b28551cf453">+21/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>configuration_test.go</strong><dd><code>Enhance tests with new DI context helpers.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/snyk/snyk-ls/pull/1302/changes#diff-e141faafb3f0507956b85e92ac3b584c7c706bd96e7f428c7ef95d0d8b2df564">+304/-185</a></td>

</tr>

<tr>
  <td><strong>notification_test.go</strong><dd><code>Add test for nil notifier panic in registerNotifier.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/snyk/snyk-ls/pull/1302/changes#diff-09b80ca28f0c9f8db60c941c24e92dfc65ded48c06249e70e2004faec30b9891">+15/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>apply_auth_config_test.go</strong><dd><code>Add tests for nil authService in endpoint change logic.</code>&nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/snyk/snyk-ls/pull/1302/changes#diff-1a300a09b5bd741eaaf57331b3a671e61e68bdd8f84a83b8d8333d497447e093">+27/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Bug fix</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>execute_command.go</strong><dd><code>Use must* helper for ErrorReporter in command handler.</code>&nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/snyk/snyk-ls/pull/1302/changes#diff-523ba95647cc7cf643285dc5a700fafb1c295bfbdc06065624232a888d1df3ef">+1/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>apply_auth_config.go</strong><dd><code>Add nil checks for authService in config apply functions.</code></dd></td>
  <td><a href="https://github.com/snyk/snyk-ls/pull/1302/changes#diff-dcbb2ac73cd481e142a8a55439036be377feb70e23eb5bfd18f59909d7678074">+12/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Additional files</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>server_test.go</strong></td>
  <td><a href="https://github.com/snyk/snyk-ls/pull/1302/changes#diff-165bad981a0cc202462a2fea7f236eab9ca3a72bd47fe8bae6b3adad801351fb">+88/-12</a>&nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: application/server/codeaction_handlers.go [hunk 1]
```diff
@@ -85,7 +84,11 @@ func GetCodeActionHandler(logger *zerolog.Logger) TextDocumentCodeActionHandler
 		default:
 		}
 
-		codeActions := di.CodeActionService().GetCodeActions(params)
+		svc, ok := codeActionServiceFromContext(ctx)
+		if !ok {
+			return nil, nil
```
Comment (rrama) [inline application/server/codeaction_handlers.go]: Probably should log a warning so if a customer asks "Where are my code actions?" we can see why they don't show.
Or is this more of a panic moment, as prod code should always have it and it forces tests to remember the service?

File: application/server/configuration.go [hunk 2]
```diff
@@ -248,8 +252,12 @@ func refreshLdxSyncOnTokenChange(ctx context.Context, conf configuration.Configu
 		return
 	}
 	logger.Info().Msg("token changed via settings, refreshing LDX-Sync configuration")
-	ldxSyncService := mustLdxSyncServiceFromContext(ctx)
-	notifier := mustNotifierFromContext(ctx)
+	ldxSyncService, ok := ldxSyncServiceFromContext(ctx)
+	if !ok {
+		logger.Warn().Msg("ldxSyncService not in context; skipping LDX-Sync refresh on token change")
```
Comment (rrama) [inline application/server/configuration.go]: Do we not want to panic to force tests to add it?

File: application/server/configuration.go [hunk 3]
```diff
@@ -298,6 +306,9 @@ func validateLockedMachineFields(settings map[string]*types.ConfigSetting, confi
 // (e.g. "Snyk Code Enabled") rather than the internal snake_case identifier.
 // See [IDE-1970].
 func notifyLockedFieldsRejected(notifier notification.Notifier, fm workflow.ConfigurationOptionsMetaData, lockedFieldGroups ...[]string) {
+	if notifier == nil {
+		return
```
Comment (rrama) [inline application/server/configuration.go]: Is it not better to updates the places that don't have a `notifier`?

File: application/server/configuration.go [hunk 4]
```diff
@@ -1010,16 +1024,16 @@ func applySnykOpenBrowserActions(conf configuration.Configuration, settings map[
 	}
 }
 
-func applyMcpConfiguration(conf configuration.Configuration, engine workflow.Engine, logger *zerolog.Logger, settings map[string]*types.ConfigSetting, triggerSource analytics.TriggerSource, configResolver types.ConfigResolverInterface) {
-	n := di.Notifier()
+func applyMcpConfiguration(ctx context.Context, conf configuration.Configuration, engine workflow.Engine, logger *zerolog.Logger, settings map[string]*types.ConfigSetting, triggerSource analytics.TriggerSource, configResolver types.ConfigResolverInterface) {
+	n, _ := notifierFromContext(ctx)
```
Comment (rrama) [inline application/server/configuration.go]: I'm not a fan of ignoring the `ok`, either use `mustNotifierFromContext` or handle the non-ok case.

File: application/server/configuration_test.go [hunk 5]
```diff
@@ -364,7 +364,10 @@ func Test_UpdateSettings(t *testing.T) {
 		// Path is init-only; apply via InitializeSettings first.
 		// TrustedFolders now goes through the settings map.
 		settingsMap[types.SettingTrustedFolders] = &types.ConfigSetting{Value: []interface{}{"trustedPath1", "trustedPath2"}, Changed: true}
-		ctx := contextWithNotifier(t.Context())
+		ctx := ctx2.NewContextWithDependencies(t.Context(), map[string]any{
+			ctx2.DepNotifier:    notification.NewMockNotifier(),
+			ctx2.DepAuthService: di.AuthenticationService(),
```
Comment (rrama) [inline application/server/configuration_test.go]: `di.TestInit` returns the deps, can you not use that?

File: application/server/notification.go [hunk 6]
```diff
@@ -154,16 +154,18 @@ func registerNotifier(conf configuration.Configuration, logger *zerolog.Logger,
...
 			l.Debug().Interface("mcpConfig", params).Msg("sending MCP config to client")
 		default:
 			l.Warn().
 				Interface("params", params).
 				Msg("received unconfigured notification object")
 		}
 	}
-	di.Notifier().CreateListener(callbackFunction)
-	l.Debug().Str("method", "registerNotifier").Msg("registered notifier")
+	if n != nil {
```
Comment (rrama) [inline application/server/notification.go]: Really feels like a case where if the notifier is nil we should bomb out. "Notifier" is in the function's name.

**Review by rrama — COMMENTED** [review]
I just really worry this is going to cause a load of silent failures which will be difficult to track down. You do add some logging when the dep isn't in the context, but not everywhere. I also worry it will make tests less realistic as if the test is missing a dep it may not test the code path it thinks it is and no one checks logs to ensure it actually is.
As such, I really don't think I can approve.

File: application/server/configuration.go (hunk 4 above)
Comment (bastiandoetsch) [reply application/server/configuration.go]: Done — replaced the `notifier, _ :=` pattern with an explicit ok-check and Warn log, matching the style already used for `ldxSyncServiceFromContext` just above it.

File: application/server/codeaction_handlers.go (hunk 1 above)
Comment (bastiandoetsch) [reply application/server/codeaction_handlers.go]: Done — added a nil guard at the top of the `GetCodeActionHandler` closure: logs a warning and returns `nil, nil` when `svc` is absent, so the missing-dep case is visible in logs rather than panicking.

File: application/server/notification.go (hunk 6 above)
Comment (bastiandoetsch) [reply application/server/notification.go]: Done — `registerNotifier` now panics on nil notifier. Since "Notifier" is in the function's name and it's always wired at server startup, nil here is a programming error — panic is the right signal and will catch it immediately in tests.

**Comment by bastiandoetsch** [comment]
Re: https://github.com/snyk/snyk-ls/pull/1302#issuecomment-4571266962

Done — `ApplyEndpointChange` now returns `false` (not `changed`) when `authService` is nil, so callers don't treat a skipped-logout as a successful endpoint switch. `ApplyAuthMethodChange` now calls `SetGlobalUser` unconditionally so the auth method persists across restarts, then guards `ConfigureProviders` behind a nil-check with a Warn log.

**Comment by bastiandoetsch** [comment]
Re: https://github.com/snyk/snyk-ls/pull/1302#issuecomment-4571463029

Done — auth state inconsistency and config-ignored findings addressed by the `apply_auth_config.go` changes above. The `addWorkspaceFolders` silent abort was already logging all missing deps by name at Error level (scanner, hoverService, scanNotifier, notifier, scanPersister, scanStateAggregator, featureFlagService, configResolver), so no additional change was needed there.

File: application/server/configuration.go (hunk 2 above)
Comment (bastiandoetsch) [reply application/server/configuration.go]: Kept Warn+return here — this runs in a hot path (token-change handler on every settings update) where a missing `ldxSyncService` should skip the LDX-Sync refresh gracefully rather than crash the server. The `must*` pattern is reserved for deps that are always wired at startup and whose absence is unambiguously a programming error; `ldxSyncService` can legitimately be absent in some deployment configurations.

File: application/server/configuration.go (hunk 3 above)
Comment (bastiandoetsch) [reply application/server/configuration.go]: Done — audited all `FromContext` call sites in `application/server/`; the remaining `notifier, _ :=` patterns now each have an explicit ok-check + Warn log. The notifier is always injected into context at server startup so it should always be present in production; the logging will surface any gap immediately rather than silently returning nil.

File: application/server/configuration_test.go (hunk 5 above)
Comment (bastiandoetsch) [reply application/server/configuration_test.go]: Done — switched the test to use `di.TestInit` for context setup instead of manually constructing the dependency map.

File: application/server/configuration.go [hunk 7]
```diff
@@ -98,7 +98,8 @@ func handlePushModel(ctx context.Context, conf configuration.Configuration, engi
 	if !conf.GetBool(types.SettingIsLspInitialized) {
 		triggerSource = analytics.TriggerSourceInitialize
 	}
-	UpdateSettings(ctx, conf, engine, logger, params.Settings, params.FolderConfigs, triggerSource, di.ConfigResolver())
+	configResolver, _ := ctx2.ConfigResolverFromContext(ctx)
```
Comment (rrama) [inline application/server/configuration.go]: Handle the `ok` or panic. Same with uses below.

File: application/server/configuration_test.go [hunk 8]
```diff
@@ -364,7 +364,13 @@ func Test_UpdateSettings(t *testing.T) {
 		// Path is init-only; apply via InitializeSettings first.
 		// TrustedFolders now goes through the settings map.
 		settingsMap[types.SettingTrustedFolders] = &types.ConfigSetting{Value: []interface{}{"trustedPath1", "trustedPath2"}, Changed: true}
-		ctx := contextWithNotifier(t.Context())
+		// Build the context from the deps returned by di.TestInit rather than manually
+		// instantiating services, so the context always uses the same instances that
+		// di.TestInit wired together.
```
Comment (rrama) [inline application/server/configuration_test.go]: Nitpick: AI comment. This is a general rule not specific information about this test.

**Review by rrama — COMMENTED** [review]
I'm still not convinced just logging instead of panicing is the way to go. I would want to pull in a 3rd person to tie break.

File: application/server/configuration.go (hunk 7 above)
Comment (bastiandoetsch) [reply application/server/configuration.go]: Done — all three sites now check the `ok` return. `handlePushModel` and `handlePullModel` return `(false, error)` when the resolver is absent; `InitializeSettings` returns `error` which `initializeHandler` propagates to the LSP client. Added `TestInitializeHandler_MissingDep_PropagatesLSPError` to verify the error reaches the client via the jrpc2 layer.

File: application/server/configuration_test.go (hunk 8 above)
Comment (bastiandoetsch) [reply application/server/configuration_test.go]: Done — removed the comment in the same commit.

File: application/server/codeaction_handlers.go [hunk 9]
```diff
@@ -60,6 +60,11 @@ func GetCodeActionHandler(logger *zerolog.Logger, svc *codeaction.CodeActionsSer
 	l := logger.With().Str("method", "CodeActionHandler").Logger()
 
 	return func(paramCtx context.Context, params types.CodeActionParams) ([]types.LSPCodeAction, error) {
+		if svc == nil {
+			l.Warn().Msg("CodeActionsService is nil; code actions are unavailable — check server startup wiring")
+			return nil, nil
```
Comment (rrama) [inline application/server/codeaction_handlers.go]: I care less about not erroring here, as I care less about the actions service, but we could do.

File: application/server/configuration.go (hunk 3 above)
Comment (rrama) [reply application/server/configuration.go]: Q for the bot: So why is this check still needed? Can it be safely dropped.

File: application/server/configuration.go [hunk 10]
```diff
@@ -378,9 +393,10 @@ func processConfigSettings(ctx context.Context, conf configuration.Configuration
 	}
 	lockedMachineFields := validateLockedMachineFields(settings, configResolver, fm, &subLogger)
 
-	applyApiEndpoints(conf, engine, logger, settings, triggerSource, configResolver)
-	applyAuthenticationMethod(conf, engine, logger, settings, triggerSource, configResolver)
-	applyToken(settings)
+	authService, _ := authenticationServiceFromContext(ctx)
```
Comment (rrama) [inline application/server/configuration.go]: Missing `ok` check.

File: application/server/server.go [hunk 11]
```diff
@@ -357,6 +429,8 @@ func mustScanPersisterFromContext(ctx context.Context) persistence.ScanSnapshotP
 	return sp
 }
 
+// scanNotifierFromContext uses the (ok) pattern; addWorkspaceFolders returns an
+// error to initializeHandler when the dep is absent.
```
Comment (rrama) [inline application/server/server.go]: Nitpick: Bad AI Comment.

File: application/server/server.go [hunk 12]
```diff
@@ -513,21 +579,23 @@ func workspaceDidChangeWorkspaceFoldersHandler(conf configuration.Configuration,
...
-				ldxSvc.RefreshConfigFromLdxSync(bgCtx, conf, engine, &logger, changedFolders, notifier)
-			}
-		}
-
-		notifier, _ := notifierFromContext(ctx)
+		// Deps below are injected by withContext if non-nil in DI wiring; each is checked
+		// before use below. HandleFolders guards nil featureFlagService internally, and a
+		// missing dep here surfaces as a no-op in HandleFolders — acceptable because
+		// workspace-folder changes are non-destructive and the handler can be retried.
+		authService, _ := authenticationServiceFromContext(ctx)
```
Comment (rrama) [inline application/server/server.go]: Use `must*` or check `ok`. Same with below.

File: domain/ide/command/apply_auth_config.go [hunk 13]
```diff
@@ -65,6 +77,12 @@ func ApplyAuthMethodChange(conf gafConfig.Configuration, authService authenticat
 		Str("new_auth_method", string(authMethod)).
 		Msg("auth method change requested")
 	types.SetGlobalUser(conf, types.SettingAuthenticationMethod, string(authMethod))
+	if authService == nil {
+		logger.Warn().
+			Str("auth_method", string(authMethod)).
+			Msg("authService is nil; auth method persisted but ConfigureProviders skipped")
```
Comment (rrama) [inline domain/ide/command/apply_auth_config.go]: Hmm, again probably fine, but could be better.

**Review by rrama — APPROVED** [review]
Recovering on panics in handlers and just throwing them as JRPC errors is nice. I know it's very not-Go, but I like it!
I am just wondering if it has any covering up issues for tests which may not check the JRPC response for errors, but it's a lot lot less of a worry now and I think it's easier to spot if a test is.

File: application/server/configuration.go (hunk 10 above)
Comment (bastiandoetsch) [reply application/server/configuration.go]: Done — changed to `mustAuthenticationServiceFromContext(ctx)` in commit `a00f20aa`.

File: application/server/server.go (hunk 12 above)
Comment (bastiandoetsch) [reply application/server/server.go]: Done — all six blank-discard patterns in `workspaceDidChangeWorkspaceFolders` replaced with `must*FromContext` in commit `a00f20aa`. The redundant nil guards were also removed.

File: application/server/configuration.go (hunk 3 above)
Comment (bastiandoetsch) [reply application/server/configuration.go]: Dropped — since all callers now use `mustNotifierFromContext`, the notifier is guaranteed non-nil before the function runs. The nil guard was dead code.

File: application/server/server.go (hunk 11 above)
Comment (bastiandoetsch) [reply application/server/server.go]: Done — removed.

File: domain/ide/command/apply_auth_config.go (hunk 13 above)
Comment (bastiandoetsch) [reply domain/ide/command/apply_auth_config.go]: Done — removed. `authService` is guaranteed non-nil by `mustAuthenticationServiceFromContext` before any handler runs; the nil guard was dead code.

**Comment by bastiandoetsch** [comment]
Re: https://github.com/snyk/snyk-ls/pull/1302#pullrequestreview-4399597479

Resolved — implemented your preferred approach: `withContext` now validates all mandatory deps and stops the server if any are nil (error returned to the LSP client via jrpc2). A `defer/recover()` also catches any synchronous panic inside handlers and converts it to a jrpc2 error + server stop. One enforcement point, one failure mode.
